xdsl/interpreters/eqsat_pdl_interp.py

```diff
@@ -83,11 +91,28 @@ class EqsatPDLInterpFunctions(PDLInterpFunctions):
     """Used for hashconsing operations. When new operations are created, if they are identical to an existing operation,
     the existing operation is reused instead of creating a new one."""
 
+    known_ops_restore_list: list[Operation] = field(default_factory=list[Operation])
+    """List of operations that have been modified during the pattern matching."""
+
     eclass_union_find: DisjointSet[eqsat.EClassOp] = field(
         default_factory=lambda: DisjointSet[eqsat.EClassOp]()
     )
     """Union-find structure tracking which e-classes are equivalent and should be merged."""
 
+    merge_list: list[MergeTodo] = field(default_factory=list[MergeTodo])
+    """List of e-classes that should be merged by `apply_matches` after the pattern matching is done."""
+
+    def modification_handler(self, op: Operation):
+        """
+        Keeps `known_ops` up to date.
+        Whenever an operation is modified, for example when its operands are updated to a different eclass value,
+        the operation is added to `known_ops_restore_list`. At the end of `apply_matches`, all the (now updated)
+        operations in `known_ops_restore_list` are put back into `known_ops`.
+        """
+        if op in self.known_ops:
+            removed = self.known_ops.pop(op)
+            self.known_ops_restore_list.append(removed)
```

Comment:
This whole approach is a bit finicky. I'm wondering if there's a better way.

Reply:
How does egg do this?

Reply:
I had a look at their implementation, egg keeps `known_ops` (called `memo` in their implementation) up to date in the rebuilding procedure (in `process_unions`).
Each parent e-node of one of the e-classes that is union-ed is put on a queue and is processed at this point.

When such a parent is processed, its children are updated to point to the union-ed e-class.
At this point, the updated e-node(/operation) is added to `memo`. The old entry is not removed but simply kept around AFAICT.

I think we should definitely have an option to rebuild like this. pdl rewritepatterns and other passes can register equalities and these can then be applied in combination with rebuilding. I’m not sure if that should be blocking for this pr though?

Reply:
Definitely not blocking for this PR!